jupyter_server/services/events/handlers.py

```diff
@@ -75,6 +79,45 @@ def on_close(self):
         self.event_bus.handlers.remove(self.logging_handler)
 
 
+class EventHandler(APIHandler):
+    """REST api handler for events"""
+
+    auth_resource = AUTH_RESOURCE
+
+    @property
+    def event_bus(self):
+        """Jupyter Server's event bus that emits structured event data."""
+        return self.settings["event_bus"]
+
+    @web.authenticated
+    @authorized
+    async def post(self):
+        payload = self.get_json_body()
+        if payload is None:
+            raise web.HTTPError(400, "No JSON data provided")
+
+        try:
+            if "timestamp" in payload:
+                timestamp = datetime.strptime(payload["timestamp"], "%Y-%m-%d %H:%M:%S")
+            else:
+                timestamp = None
+
+            self.event_bus.record_event(
+                schema_name=payload["schema_name"],
+                version=payload["version"],
+                event=payload["event"],
+                timestamp_override=timestamp,
+            )
+            self.set_status(204)
+            self.finish()
+        except KeyError as ke:
+            prop = str(ke).split(":")[-1].strip()
```

Comment:
Hm, I'm surprised this raises a `KeyError` and not a `ValidationError` from jsonschema. What exactly would raise this `KeyError`? It seems that missing keys would be detected by the validator. 

I'm a bit concerned about the dependency on an error message here. Is this message structured in a way that we can rely on long term?

Reply:
`KeyError` would happen for any missing key in the payload, for example payload['version'] if 'version' is not present. `record_event` expects `schema_name`, `version`, and `event` as required arguments, so this was my attempt to shorten the code to validate request parameters. I agree with your concern about parsing the missing key, long term if the `KeyError` formatting changes, it will likely break this code.

Here is the proposed change, which checks for each prop individually:
```py
if 'schema_name' not in payload:
    raise web.HTTPError(400, f"'schema_name' missing in JSON data")

if 'version' not in payload:
    raise web.HTTPError(400, f"'version' missing in JSON data")

if 'event' not in payload:
    raise web.HTTPError(400, f"'event' missing in JSON data")

# call record_event
```
I was hoping there was a better way to validate request parameters, let me know if there is an alternative.

Reply:
I'm thinking we should add a `validate_model` function that validates the REST API model for requests before doing any further actions. This ensures that the income request is valid before we do anything else. 

I'm imagining something like below that lives outside the handler: 
```python
def validate_model(data):
    """Validate the REST API event model."""
    required_keys = {"schema_name", "version", "event"}
    for key in required_keys:
        if key not in payload:
            raise web.HTTPError(400, f"Missing `{key}` in the JSON request body.")
```

Reply:
Done!